### PR TITLE
Feature/401 api call to retrieve public beneficial owner data

### PR DESCRIPTION
**Pull request by azafarCH**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-401?search_id=12dc7f50-79e9-4e27-8661-e4344e2f7a01

### Change description
Create call in FilingsService that returns a list of Pscs (persons of siginificant control)

### Work checklist

- [x] Tests added where applicable

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    private CompanyProfileApi getCompanyProfile(String companyNumber, String passThroughTokenHeader) throws ServiceException {
+        try {
+            CompanyProfileApi companyProfileData = apiClientService
+                    .getOauthAuthenticatedClient(passThroughTokenHeader)
+                    .company()
+                    .get("/company/" + companyNumber)
+                    .execute()
+                    .getData();
+
+            var logMap = new HashMap<String, Object>();
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java:41]: Log before doing the retrieve

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+            PscsApi pscsData = pscsApiResponse.getData();
+
+            var logMap = new HashMap<String, Object>();
+
+            if (pscsApiResponse.getStatusCode() == HttpServletResponse.SC_NOT_FOUND) {
+                logMap.put(COMPANY_NUMBER, companyNumber);
+                ApiLogger.info("No PSCs for Company Number " + logMap);
+            } else {
+                logMap.put(PSCS_NAME, pscsData.getItems().get(0).getName());
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java:71]: Use companyNumber in log here too.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                ApiLogger.info("Retrieving PSCs data for PSC name " + logMap);
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java:72]: Don't just log the first PSC. Loop through all and debug log each. But not good to ever log personal info so better to remove the logging of the individual(s). Maybe just log the length of list returned. Please coordinate with Sifat.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java [hunk 4]
```diff
@@ -79,7 +88,7 @@ private OfficersApi getOfficers(String companyNumber, String passThroughTokenHea
             if (officersApiResponse.getStatusCode() == HttpServletResponse.SC_NOT_FOUND) {
                 ApiLogger.info("No Officer for Company Number ", logMap);
             } else {
-                for (CompanyOfficerApi officer: officersData.getItems()) {
+                for (CompanyOfficerApi officer : officersData.getItems()) {
```
Comment (sjahir) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java:91]: remove space

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java [hunk 5]
```diff
@@ -92,4 +101,36 @@ private OfficersApi getOfficers(String companyNumber, String passThroughTokenHea
             throw new ServiceException(e.getMessage(), e);
         }
     }
+
+    private PscsApi getPSCs(String companyNumber, String passthroughTokenHeader) throws ServiceException{
+        var logMap = new HashMap<String, Object>();
+        logMap.put(COMPANY_NUMBER, companyNumber);
```
Comment (sjahir) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java:107]: add a space before line 107

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java [hunk 6]
```diff
@@ -32,13 +36,18 @@ public OfficersApi getOfficerApi() {
         return officers;
     }
 
+    public PscsApi getPscsApi() {
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java:39]: remove Api from method name

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java [hunk 7]
```diff
@@ -92,4 +101,36 @@ private OfficersApi getOfficers(String companyNumber, String passThroughTokenHea
...
+                    .execute();
+
+            PscsApi pscsData = pscsApiResponse.getData();
+
+            if (pscsApiResponse.getStatusCode() == HttpServletResponse.SC_NOT_FOUND) {
+                logMap.put(COMPANY_NUMBER, companyNumber);
+                ApiLogger.info("No PSCs for Company Number " + logMap);
+            } else {
+                for(PscApi psc :  pscsData.getItems()) {
+                    logMap.put(PSCS_NAME, psc.getName());
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java:124]: move put out of for loop as only needs to be done once.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalServiceTest.java [hunk 8]
```diff
@@ -75,40 +88,94 @@ void testInitialisePublicDataIsSuccessful() throws IOException, URIValidationExc
         var companyProfileApi = new CompanyProfileApi();
         var officersApi = new OfficersApi();
         var companyOfficerApi = new CompanyOfficerApi();
+        var pscsApi = new PscsApi();
+        var companyPSCsApi = new PscApi();
```
Comment (AndreKramer7654) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalServiceTest.java:92]: why two vars that are the same thing?

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java (hunk 3 above)
Comment (azafarCH) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java]: I used ApiLogger.debug here instead now, as was done with Sifat's ticket is that sufficient?

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalServiceTest.java (hunk 8 above)
Comment (azafarCH) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalServiceTest.java:92]: I can mock it or change the name but that would affect the consistency with the officers ones, is that fine or is it better to leave it as it is?

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java (hunk 6 above)
Comment (azafarCH) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java:39]: Should the getOfficerApi() also be changed to getOfficerApi() ?

Comment (AndreKramer7654) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java:39]: I just took the standard pattern for getters. They just need to use the same convention I'd say (getCompanyProfile is different). I'd prefer no Api

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalServiceTest.java (hunk 8 above)
Comment (AndreKramer7654) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalServiceTest.java:92]: It's minor leave as is.

**Review by AndreKramer7654 — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java [hunk 9]
```diff
@@ -92,4 +100,34 @@ private OfficersApi getOfficers(String companyNumber, String passThroughTokenHea
...
+            ApiResponse<PscsApi> pscsApiResponse = apiClientService
+                    .getOauthAuthenticatedClient(passthroughTokenHeader)
+                    .pscs()
+                    .list("/company/" + companyNumber + "/persons-with-significant-control")
+                    .execute();
+
+            PscsApi pscsData = pscsApiResponse.getData();
+
+            if (pscsApiResponse.getStatusCode() == HttpServletResponse.SC_NOT_FOUND) {
+                ApiLogger.info("No PSCs for Company Number " + logMap);
```
Comment (bantao-ch) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PublicDataRetrievalService.java:120]: two args needed here -- "," not "+"

**Review by bantao-ch — CHANGES_REQUESTED** [review]
See comment in logger --

**Review by bantao-ch — APPROVED** [review]

**Review by jkayyala — APPROVED** [review]

**Pull request merged**